Sources/SwiftInspectorCore/TypeLocationAnalyzer.swift

```diff
@@ -0,0 +1,174 @@
+// Created by Michael Bachand on 3/28/20.
+//
+// Distributed under the MIT License
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in all
+// copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
+// SOFTWARE.
+
+import Foundation
+import SwiftSyntax
+
+// MARK: - TypeLocationAnalyzer
+
+public final class TypeLocationAnalyzer: Analyzer {
+
+  /// - Parameter typeName: The name of the type to locate.
+  /// - Parameter cachedSyntaxTree: The cached syntax tree to return the AST tree from
+  public init(typeName: String, cachedSyntaxTree: CachedSyntaxTree = .init()) {
+    self.typeName = typeName
+    self.cachedSyntaxTree = cachedSyntaxTree
+  }
+
+  /// Analyzes the imports of the Swift file
+  /// - Parameter fileURL: The fileURL where the Swift file is located
+  public func analyze(fileURL: URL) throws -> [LocatedType] {
+    let syntax: SourceFileSyntax = try cachedSyntaxTree.syntaxTree(for: fileURL)
+    var result = [LocatedType]()
+    let reader = TypeLocationSyntaxReader() { [unowned self] locatedType in
+      guard self.typeName == locatedType.name else { return }
+      result.append(locatedType)
+    }
+    _ = reader.visit(syntax)
+
+    return result
+  }
+
+  // MARK: Private
+
+  private let cachedSyntaxTree: CachedSyntaxTree
+  private let typeName: String
+}
+
+// TODO: Update to use SyntaxVisitor when this bug is resolved (https://bugs.swift.org/browse/SR-11591)
+private final class TypeLocationSyntaxReader: SyntaxRewriter {
+  init(onNodeVisit: @escaping (_ locatedType: LocatedType) -> Void) {
+    self.onNodeVisit = onNodeVisit
+  }
+
+  override func visitAny(_ node: Syntax) -> Syntax? {
```

Comment:
Is this more efficient than using `  override public func visit(_ node: ClassDeclSyntax) -> DeclSyntax {` and similar methods for the different types?

Reply:
@joeyg I started there and then ended up here since it initially made the code a bit easier to follow. Then I made `anyLocatedType`, which I think would allow me to go back to those methods. My decisionmaking wasn't around efficiency.

Reply:
From clicking into the `SyntaxRewriter` code it didn't feel like there is a ton of efficiency to gain in that respect. I think where there is more to gain is bailing out of things early or short circuiting the visitations, but I haven't figured out if/how that is possible.

Reply:
Yeah, using SyntaxRewriter here won't give you that, but we can move this to use SyntaxVisitor by using:


```swift
let syntax = try cachedSyntaxTree.syntaxTree(for: fileURL)

var reader = TypeLocationSyntaxReader(shouldVisitIdentifier: shouldVisit) { [unowned self] node in
    ...
}
syntax.walk(&reader)

func shouldVisit(identifier: String) -> Bool {
  ...
}

private final class TypeLocationSyntaxReader: SyntaxVisitor {
  init(
    shouldVisitIdentifier: @escaping (String) -> Bool,
    onNodeVisit: @escaping (SomeNode) -> Void
  )
  {
    self.shouldVisitIdentifier = shouldVisitIdentifier
    self.onNodeVisit = onNodeVisit
  }

  func visit(_ node: ClassDeclSyntax) -> SyntaxVisitorContinueKind {
    shouldVisitIdentifier(node.identifier.text) ? .visitChildren : .skipChildren
  }

  func visit(_ node: StructDeclSyntax) -> SyntaxVisitorContinueKind {
    shouldVisitIdentifier(node.identifier.text) ? .visitChildren : .skipChildren
  }

  func visit(_ node: EnumDeclSyntax) -> SyntaxVisitorContinueKind {
    shouldVisitIdentifier(node.identifier.text) ? .visitChildren : .skipChildren
  }

  func visit(_ node: SomeNode) -> SyntaxVisitorContinueKind {
    onNodeVisit(node)
    return .visitChildren
  }

  let shouldVisitIdentifier: (String) -> Bool
  let onNodeVisit: (SomeNode) -> Void
}
```

Reply:
I recently discovered the walk function from @joeyg FWIW 😉

Reply:
Cool I'll try this. I plan to consider visiting "any" node unless there is feedback that I shouldn't.

Reply:
@bachand it's probably more performant to not do that and just do processing. in the nodes you care about. That way we need to only move from SwiftSyntax to SwiftInspector for the nodes we care about instead of checking twice for each node type (once in SwiftInspector and once in SwiftSyntax)

Reply:
@fdiaz I will give it a try and see where I get

Reply:
I moved to `SyntaxVisitor`. I didn't see an easy way to skip children since I need to count the tokens within nodes to determine the current line number.